### PR TITLE
Terminate ordered lucene collector on cancel/kill 

### PR DESCRIPTION
**Pull request by seut**
If an ordered lucene collector is running inside a child thread, it should stop processing and the thread should terminate if cancelled/killed.
Otherwise it may still occupy resources (e.g. account memory) AFTER the job was killed and shared resources (ram accounting) are released.

Supersedes #9628.

### CODE REVIEW

**Comment by seut** [comment]
@mfussenegger I've pushed a new fixup which uses the Killable interface and changes some signatures to CompletionStage as discussed. Please check this draft.

File: shared/src/main/java/io/crate/concurrent/CompletableFutures.java [hunk 1]
```diff
@@ -57,4 +67,349 @@ private CompletableFutures() {
         }
     }
 
+    public static <T> CompletionStage<T> asKillable(CompletableFuture<T> delegate,
+                                                    Consumer<Throwable> beforeExceptional) {
```
Comment (mfussenegger) [inline shared/src/main/java/io/crate/concurrent/CompletableFutures.java]:
```suggestion
                                                    Consumer<Throwable> onKill) {
```

File: shared/src/main/java/io/crate/concurrent/CompletableFutures.java [hunk 2]
```diff
@@ -57,4 +67,349 @@ private CompletableFutures() {
...
+        return new KillableCompletableFuture<>(delegate, beforeExceptional);
+    }
+
+    @SuppressForbidden
```
Comment (mfussenegger) [inline shared/src/main/java/io/crate/concurrent/CompletableFutures.java]: Just curious, what's the forbidden part here?

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java [hunk 3]
```diff
@@ -72,6 +79,11 @@
     private int batchSize;
     private boolean batchSizeReduced = false;
 
+    @Nullable
+    private volatile KillableCollector currentCollector;
+
+    @Nullable
+    private volatile Throwable killed;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: I am not sure if we need both of these to be volatile.

We could probably remove the check here https://github.com/crate/crate/pull/9650/files#diff-1360f05e99db1b25d5f53dec74c7abb2R137 and always depend on the kill raising via the Collector. I think the operations happening between those actions are cheap enough to ignore.

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java [hunk 4]
```diff
@@ -205,4 +231,83 @@ private Query query(FieldDoc lastDoc) {
...
+        }
+
+        @Override
+        public void kill(@Nonnull Throwable t) {
+            if (killed != null) {
+                return;
+            }
+            killed = t;
+            var leafCollector = currentLeafCollector;
+            if (leafCollector != null) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: I think there is a race condition between getLeafCollector and kill:

```
t1: Is before `killed = t;`
t2: Is past `raiseIfKilled(killed);` but hasn`t done the assignment to `currentLeafCollector`
```

In that case the kill would be swallowed I think. Could probably be solved with a compare and set operation

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/OrderedLuceneBatchIteratorFactory.java [hunk 5]
```diff
@@ -102,22 +104,24 @@
             );
         }
 
-        private CompletableFuture<List<KeyIterable<ShardId, Row>>> tryFetchMore(ShardId shardId) {
+        private CompletionStage<List<KeyIterable<ShardId, Row>>> tryFetchMore(ShardId shardId) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/engine/collect/collectors/OrderedLuceneBatchIteratorFactory.java]: I think here we could actually return the new `KillableCompletableFuture`. Otherwise the call-site has to know that this might implement `Killable`, that is not very common.

By changing the interface to return `CompletionStage` I was referring to `loadNextBatch`, but that could probably be a dedicated PR.

Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/collect/collectors/OrderedLuceneBatchIteratorFactory.java]: Or maybe we expose a `KillableCompletionStage` here?

**Comment by seut** [comment]
hm actually this is a draft, my intention for feedback was more on a general level (was not ready for proper review).

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/OrderedLuceneBatchIteratorFactory.java (hunk 5 above)
Comment (seut) [reply sql/src/main/java/io/crate/execution/engine/collect/collectors/OrderedLuceneBatchIteratorFactory.java]: > By changing the interface to return CompletionStage I was referring to loadNextBatch

That's already in place.

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java (hunk 4 above)
Comment (seut) [reply sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: I don't think so as `killed` is passed to the leaf collector ctor.

File: shared/src/main/java/io/crate/concurrent/CompletableFutures.java (hunk 2 above)
Comment (seut) [reply shared/src/main/java/io/crate/concurrent/CompletableFutures.java]: See https://github.com/crate/crate/blob/master/gradle/forbidden-signatures.txt#L5

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java (hunk 3 above)
Comment (seut) [reply sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: > I am not sure if we need both of these to be volatile.

Both can be set and used from different threads.

```
    t1                       t2
                          OrderedDocCollector
                          -> collect()
                            -> set currentCollector
                            -> check killed and raise

  kill()
      ------------------> OrderedDocCollector
                        -> set killed
                        -> check currentCollector and call kill()  if set
```

Comment (seut) [reply sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: > We could probably remove the check here https://github.com/crate/crate/pull/9650/files#diff-1360f05e99db1b25d5f53dec74c7abb2R137 and always depend on the kill raising via the Collector. I think the operations happening between those actions are cheap enough to ignore.

Yep agree.

**Comment by seut** [comment]
@mfussenegger So I treat your feedback as "in-general ok" and will shape this draft into a proper PR.

**Comment by mfussenegger** [comment]
yep, general pattern looks okay.

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java (hunk 4 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: Yes, but only if `killed` has been assigned before the leaf collector is created.

The scenario I have in mind is that `killed` hasn't been assignd to yet, and the other thread read the volatile (unassigned) 'killed' but didn't finish the 'currentLeafCollector'  assignment yet.

How about having a final AtomicReference in the top most class and the collector + leaf collector receive a reference to that? That would replace the individual killed attributes.

Comment (seut) [reply sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: Right, there is still a race condition, good catch.

> How about having a final AtomicReference in the top most class and the collector + leaf collector receive a reference to that? That would replace the individual killed attributes.

Yep, sounds good.

Comment (seut) [reply sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: Pushed a fixup.

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java [hunk 6]
```diff
@@ -231,82 +222,55 @@ private Query query(FieldDoc lastDoc) {
...
+        private final Consumer<Void> raiseIfKilled;
 
-        @Nullable
-        private volatile KillableLeafCollector currentLeafCollector;
-
-        @Nullable
-        private volatile Throwable killed;
-
-        public KillableCollector(Collector delegate, @Nullable Throwable killed) {
+        public KillableCollector(Collector delegate, Consumer<Void> raiseIfKilled) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]:
```suggestion
        public KillableCollector(Collector delegate, Runnable raiseIfKilled) {
```
